tapgarden/caretaker.go

```diff
@@ -888,65 +898,105 @@ func (b *BatchCaretaker) stateStep(currentState BatchState) (BatchState, error)
-				// where it was created at and the script key
-				// for that asset.
-				baseKey := universe.BaseKey{
-					MintingOutpoint: wire.OutPoint{
-						Hash:  confInfo.Tx.TxHash(),
-						Index: b.anchorOutputIndex,
-					},
-					ScriptKey: &newAsset.ScriptKey,
-				}
-
-				// With both of those assembled, we can now
-				// register issuance which takes the amount and
-				// proof of the minting event.
-				uniGen := universe.GenesisWithGroup{
-					Genesis: newAsset.Genesis,
-				}
-				if groupKey != nil {
-					uniGen.GroupKey = groupKey
-				}
-				mintingLeaf := &universe.MintingLeaf{
-					GenesisWithGroup: uniGen,
-
-					// The universe tree store only the
-					// asset state transition and not also
-					// the proof file checksum (as the root
-					// is effectively a checksum), so we'll
-					// use just the state transition.
-					GenesisProof: mintingProof,
-					Amt:          newAsset.Amount,
-				}
-				_, err = b.cfg.Universe.RegisterIssuance(
-					ctx, uniID, baseKey, mintingLeaf,
-				)
-				if err != nil {
-					return 0, fmt.Errorf("unable to "+
-						"register issuance: %v", err)
-				}
+			// local base universe. We skip this step if there is
+			// no universe configured.
+			if b.cfg.Universe == nil {
+				continue
+			}
+
+			// The universe ID serves to identifier the universe
+			// root we want to add this asset to. This is either the
+			// assetID or the group key.
+			uniID := universe.Identifier{
+				AssetID: assetID,
+			}
+
+			groupKey := newAsset.GroupKey
+			if groupKey != nil {
+				uniID.GroupKey = &groupKey.GroupPubKey
+			}
+
+			log.Debugf("Preparing asset for registration with "+
+				"universe, key=%v", spew.Sdump(uniID))
+
+			// The base key is the set of bytes that keys into the
+			// universe, this'll be the outpoint where it was
+			// created at and the script key for that asset.
+			baseKey := universe.BaseKey{
+				MintingOutpoint: wire.OutPoint{
+					Hash:  confInfo.Tx.TxHash(),
+					Index: b.anchorOutputIndex,
+				},
+				ScriptKey: &newAsset.ScriptKey,
+			}
+
+			// With both of those assembled, we can now register
+			// issuance which takes the amount and proof of the
+			// minting event.
+			uniGen := universe.GenesisWithGroup{
+				Genesis: newAsset.Genesis,
+			}
+			if groupKey != nil {
+				uniGen.GroupKey = groupKey
+			}
+			mintingLeaf := &universe.MintingLeaf{
+				GenesisWithGroup: uniGen,
+
+				// The universe tree store only the asset state
+				// transition and not also the proof file
+				// checksum (as the root is effectively a
+				// checksum), so we'll use just the state
+				// transition.
+				GenesisProof: mintingProof,
+				Amt:          newAsset.Amount,
+			}
+			universeItems <- &universe.IssuanceItem{
+				ID:   uniID,
+				Key:  baseKey,
+				Leaf: mintingLeaf,
+			}
+		}
+		close(universeItems)
+
+		// Batch insert the new proofs into the local universe, if we
+		// have one configured.
+		if b.cfg.Universe != nil {
```

Comment:
If we don't have a universe configured, then we also don't need `universeItems` either. Though I think this is really only the case for unit tests (avoiding a nil pointer painc).